test/protocol/break_test.rb

```diff
@@ -5,43 +5,93 @@
 module DEBUGGER__
   class BreakTest1 < ProtocolTestCase
     PROGRAM = <<~RUBY
-      1| module Foo
-      2|   class Bar
-      3|     def self.a
-      4|       "hello"
-      5|     end
-      6|   end
-      7|   Bar.a
-      8|   bar = Bar.new
-      9| end
+     1| class Foo
+     2|   def self.bar
+     3|     "hello"
+     4|   end
+     5|
+     6|   def baz
+     7|     10
+     8|   end
+     9| end
+    10|
+    11| Foo.bar
+    12| f = Foo.new
+    13| f.baz
     RUBY
 
-    def test_break_stops_at_correct_place
+    def test_set_breakpoints_sets_line_breakpoints
       run_protocol_scenario PROGRAM do
-        req_add_breakpoint 5
+        req_add_breakpoint 4
         req_continue
-        assert_line_num 5
+        assert_line_num 4
 
         assert_locals_result(
           [
-            { name: "%self", value: "Foo::Bar", type: "Class" },
+            { name: "%self", value: "Foo", type: "Class" },
             { name: "_return", value: "hello", type: "String" }
           ]
         )
 
-        req_add_breakpoint 9
+        req_add_breakpoint 13
         req_continue
-        assert_line_num 9
+        assert_line_num 13
 
         assert_locals_result(
           [
-            { name: "%self", value: "Foo", type: "Module" },
-            { name: "bar", value: /#<Foo::Bar/, type: "Foo::Bar" }
+            { name: "%self", value: "main", type: "Object" },
+            { name: "f", value: /#<Foo.*>/, type: "Foo" }
           ]
         )
         req_terminate_debuggee
       end
     end
+
+    def test_set_function_breakpoints_sets_instance_method_breakpoints
+      run_protocol_scenario PROGRAM, cdp: false do
+        res_set_function_breakpoints([{ name: "Foo.bar" }])
+        req_continue
+        assert_line_num 3
+
+        assert_locals_result(
+          [
+            { name: "%self", value: "Foo", type: "Class" }
+          ]
+        )
+
+        req_continue
+      end
+    end
+
+    def test_set_function_breakpoints_sets_breakpoint_with_condition
+      run_protocol_scenario PROGRAM, cdp: false do
+        res_set_function_breakpoints([{ name: "Foo.bar", condition: "$foo == 1" }])
+        req_continue
+      end
+    end
+
+    def test_set_function_breakpoints_sets_class_method_breakpoints
+      run_protocol_scenario PROGRAM, cdp: false do
+        req_add_breakpoint 13
+        req_continue
+        assert_line_num 13
+
+        res_set_function_breakpoints([{ name: "f.baz" }])
+        req_continue
+
+        assert_line_num 7
+
+        req_terminate_debuggee
+      end
+    end
+
+    def test_set_function_breakpoints_unsets_method_breakpoints
+      run_protocol_scenario PROGRAM, cdp: false do
+        res_set_function_breakpoints([{ name: "Foo::Bar.a" }])
```

Comment:
This breakpoint wouldn't have triggered anyways since it doesn't match anything in the code.